### PR TITLE
[SqlAzureDacpacDeploymentV1] Support for Service Princpal authentication

### PR DESCRIPTION
**Pull request by hangar18rip**
**Task name**: SqlAzureDacpacDeploymentV1

**Description**: Implementation proposal for SP support

**Documentation changes required:** Y

**Added unit tests:** Y

**Attached related issue:** Y : #13117

**Checklist**:
- [x] Task version was bumped - please check [instruction](https://github.com/microsoft/azure-pipelines-tasks/tree/master/docs/taskversionbumping.md) how to do it
- [x] Checked that applied changes work as expected

### CODE REVIEW

**Comment by odeatavanade** [comment]
Hello, is it possible to get details on pipeline failures ?

**Comment by hangar18rip** [comment]
Hello, is it possible to get details on pipeline failures please ?

**Comment by hangar18rip** [comment]
hello @chshrikh / @Ajay-MS and @nadesu : some help would be appreciated to fix the issues with the pipeline failures. Any clue ?

**Comment by nadesu** [comment]
@AmrutaKawade pls look into this PR

File: Tasks/SqlAzureDacpacDeploymentV1/SqlAzureActions.ps1
Comment (AmrutaKawade) [inline Tasks/SqlAzureDacpacDeploymentV1/SqlAzureActions.ps1:183]: Remove $endpoint its not being used

File: Tasks/SqlAzureDacpacDeploymentV1/SqlAzureActions.ps1
Comment (AmrutaKawade) [inline Tasks/SqlAzureDacpacDeploymentV1/SqlAzureActions.ps1:351]: Need to add token here as well. Otherwise it will fail at this step

File: Tasks/SqlAzureDacpacDeploymentV1/Utility.ps1
Comment (AmrutaKawade) [inline Tasks/SqlAzureDacpacDeploymentV1/Utility.ps1:202]: add error message in resources.resjson. Use like this `throw  (Get-VstsLocString`

**Review by AmrutaKawade — CHANGES_REQUESTED** [review]
> hello @chshrikh / @Ajay-MS and @nadesu : some help would be appreciated to fix the issues with the pipeline failures. Any clue ?

What help do you need

**Comment by AmrutaKawade** [comment]
Update documentation as well. Whether service principal needs any permissions on sql server before using?

**Review by AmrutaKawade — COMMENTED** [review]
Update documentation as well. Whether service principal needs any permissions on sql server before using?

**Comment by hangar18rip** [comment]
I will have a look at your comments this afternoon

File: Tasks/SqlAzureDacpacDeploymentV1/Utility.ps1
Comment (hangar18rip) [reply Tasks/SqlAzureDacpacDeploymentV1/Utility.ps1]: Added ```SAD_NoServerSpecified```

File: Tasks/SqlAzureDacpacDeploymentV1/SqlAzureActions.ps1
Comment (hangar18rip) [reply Tasks/SqlAzureDacpacDeploymentV1/SqlAzureActions.ps1:351]: added where needed

**Comment by hangar18rip** [comment]
> > hello @chshrikh / @Ajay-MS and @nadesu : some help would be appreciated to fix the issues with the pipeline failures. Any clue ?
> 
> What help do you need?

Can I get details on the failure of CI pipelines please ?

**Comment by AmrutaKawade** [comment]
> > > hello @chshrikh / @Ajay-MS and @nadesu : some help would be appreciated to fix the issues with the pipeline failures. Any clue ?
> > 
> > 
> > What help do you need?
> 
> Can I get details on the failure of CI pipelines please ?

As you have changed common file
Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1

The following tasks should have their versions bumped due to changes in common: AzureFileCopyV2, AzureFileCopyV3, AzureFileCopyV4

**Comment by hangar18rip** [comment]
hi @asranja / @AmrutaKawade. We don't get Azure Pipeline outputs. Can you tell me more about the errors please ?

**Comment by AmrutaKawade** [comment]
You missed to bump version of AzureFileCopyV4. @asranja Can you also review PR once?

**Comment by hangar18rip** [comment]
> hi @asranja / @AmrutaKawade. We don't get Azure Pipeline outputs. Can you tell me more about the errors please ?

I did in commit [9e507d5](https://github.com/microsoft/azure-pipelines-tasks/pull/13770/commits/9e507d5b15f5d13898eb04adcb3e228b21400ed0). I will bump it once more in a minute

**Comment by hangar18rip** [comment]
> hi @asranja / @AmrutaKawade. We don't get Azure Pipeline outputs. Can you tell me more about the errors please ?

I did in commit [9e507d5](https://github.com/microsoft/azure-pipelines-tasks/pull/13770/commits/9e507d5b15f5d13898eb04adcb3e228b21400ed0). I will bump it another time

**Comment by sujitks** [comment]
Hi @chshrikh @asranja @AmrutaKawade @nadesu, we thought SPN based authentication is already there and now realized it was not. Thanks to @hangar18rip for this if it is working. 

We need this for our deployment of DACPAC to configure custom users (AAD APP). I hope you guys get time to review and respond quickly.

**Review by sujitks — APPROVED** [review]
Looks good to me

**Comment by hangar18rip** [comment]
Hi @asranja and @AmrutaKawade. Is there anything blocking this PR I can solve on my side ?

**Review by AmrutaKawade — APPROVED** [review]
I am good with this PR. I want @asranja to have to look once as he is one of the initial author of task.

File: Tasks/SqlAzureDacpacDeploymentV1/Utility.ps1
Comment (asranja) [inline Tasks/SqlAzureDacpacDeploymentV1/Utility.ps1:295]: Please add the condition to return servicePrincipal authentication type.

File: Tasks/SqlAzureDacpacDeploymentV1/Utility.ps1
Comment (asranja) [inline Tasks/SqlAzureDacpacDeploymentV1/Utility.ps1:109]: Set the value of $token to be **** when isOutputSecure flag is set, otherwise it will leak secrets into the logs.

File: Tasks/SqlAzureDacpacDeploymentV1/Strings/resources.resjson/en-US/resources.resjson
Comment (asranja) [inline Tasks/SqlAzureDacpacDeploymentV1/Strings/resources.resjson/en-US/resources.resjson:81]: we can be explicit here saying that no source or target server was specified.

File: Tasks/SqlAzureDacpacDeploymentV1/SqlAzureActions.ps1
Comment (asranja) [inline Tasks/SqlAzureDacpacDeploymentV1/SqlAzureActions.ps1:302]: comment required?

File: Tasks/SqlAzureDacpacDeploymentV1/Utility.ps1
Comment (hangar18rip) [reply Tasks/SqlAzureDacpacDeploymentV1/Utility.ps1:295]: Good point but which condition makes the diference between aadAuthenticationIntegrated and servicePrincipal ?

File: Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1
Comment (asranja) [inline Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1:273]: you would also need to add overrideResourceType to `Get-SpnAccessTokenUsingCertificate` and `Get-MsiAccessToken` methods as well.

File: Tasks/SqlAzureDacpacDeploymentV1/DeploySqlAzure.ps1
Comment (asranja) [inline Tasks/SqlAzureDacpacDeploymentV1/DeploySqlAzure.ps1:101]: Instead of hardcoding `https://database.windows.net/`, get the value from `$endpoint.Data.sqlDatabaseDnsSuffix`. The output is in form `.database.windows.net`. You will have to remove dot(.) from the start and add https.

This is needed because the task also targets other national clouds, so this value will be different for different clouds.

Comment (hangar18rip) [reply Tasks/SqlAzureDacpacDeploymentV1/DeploySqlAzure.ps1:101]: I will investigate all those points by the week-end. Need time to

File: Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1
Comment (hangar18rip) [reply Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1]: this [commit](https://github.com/microsoft/azure-pipelines-tasks/pull/13770/commits/33b987a9d3b50e5ef74cd193ddd33f69f62ec2c2) achieves this. Deployed a copy of the task and tested it live

File: Tasks/SqlAzureDacpacDeploymentV1/task.json
Comment (asranja) [inline Tasks/SqlAzureDacpacDeploymentV1/task.json:97]: helpMarkDown needs to be updated for Service Principal info.

**Review by asranja — APPROVED** [review]

**Comment by asranja** [comment]
@hangar18rip Thanks for taking this change. Seems fine to me overall. Can you build all tasks whose task.json you have updated once before merging. This will auto update the task.loc and resource.resjon files. `node make.js build --task TaskName`

**Comment by hangar18rip** [comment]
> @hangar18rip Thanks for taking this change. Seems fine to me overall. Can you build all tasks whose task.json you have updated once before merging. This will auto update the task.loc and resource.resjon files. `node make.js build --task TaskName`

done ;)

**Comment by hangar18rip** [comment]
hi @nadesu @chshrikh, how can we continue on with this PR?

**Comment by nadesu** [comment]
@hangar18rip we are on festival holiday in India. Will do the needful early next week. Cheers!

**Comment by hangar18rip** [comment]
> @hangar18rip we are on festival holiday in India. Will do the needful early next week. Cheers!

Sorry. I was not aware of that. See you next week

**Comment by hangar18rip** [comment]
hi @asranja. Is there a way to know when tasks are getting deployed to production ? the PR has been merged 20 days ago

**Pull request merged**
